server/src/test/java/org/elasticsearch/cluster/routing/AllocationIdIT.java

```diff
@@ -0,0 +1,298 @@
+         *  -> node0 becomes a new primary with the same historyUUID if (*) has a real allocation id
+         *  -> node0 has a RED index if (*) points to a fake shard -> node requires another AllocateStalePrimary
+         * - index same amount of docs to node1 as it was added at (2)
+         *
+         * - node1 and node2 have the same number of docs ( but with different docs )
+         * - bring node2 back
+         * -> no recovery take place if a shard allocation id at (*) is persisted => nodes are fully in-sync but have diff docs
+         * -> due to fake allocation id at (*) AllocateStalePrimary is forced for the 2nd time and a full recovery takes place
+         */
+        // initial set up
+        final String indexName = "index42";
+        String node0 = internalCluster().startMasterOnlyNode();
+        String node1 = internalCluster().startNode();
+        createIndex(indexName);
+        final int numDocs = indexDocs(indexName);
+        final IndexSettings indexSettings = getIndexSettings(indexName, node1);
+        final String primaryNodeId = getNodeIdByName(node1);
+        final Set<String> allocationIds = getAllocationIds(indexName);
+        final ShardId shardId = new ShardId(resolveIndex(indexName), 0);
+        final Path indexPath = getIndexPath(node1, shardId);
+        assertThat(allocationIds, hasSize(1));
+        final Set<String> historyUUIDs = historyUUIDs(node1, indexName);
+        String node2 = internalCluster().startNode();
+        ensureGreen(indexName);
+        assertSameDocIdsOnShards();
+        // initial set up is done
+
+        internalCluster().stopRandomNode(InternalTestCluster.nameFilter(node1));
+
+        // index more docs to node2 (formal replica)
+        int numExtraDocs = randomIntBetween(10, 100);
+        {
+            IndexRequestBuilder[] builders = new IndexRequestBuilder[numExtraDocs];
+            for (int i = 0; i < builders.length; i++) {
+                builders[i] = client().prepareIndex(indexName, "type").setSource("foo", "bar2");
+            }
+
+            indexRandom(true, false, false, Arrays.asList(builders));
+            flush(indexName);
+            assertHitCount(client(node2).prepareSearch(indexName).setQuery(matchAllQuery()).get(), numDocs + numExtraDocs);
+        }
+
+        internalCluster().stopRandomNode(InternalTestCluster.nameFilter(node2));
+
+        // create fake corrupted marker
+        putFakeCorruptionMarker(indexSettings, shardId, indexPath);
+
+        // thanks to master node1 is out of sync
+        node1 = internalCluster().startNode();
+
+        // there is only _stale_ primary
+        checkNoValidShardCopy(indexName, shardId);
+
+        // allocate stale primary
+        client(node1).admin().cluster().prepareReroute()
+            .add(new AllocateStalePrimaryAllocationCommand(indexName, 0, primaryNodeId, true))
+            .get();
+
+        internalCluster().stopRandomNode(InternalTestCluster.nameFilter(node1));
```

Comment:
why do we need to stop the node? can't we keep it alive and just remove the marker?